Do less unnecessary work

marienz opened the pull request:
This PR tries to keep sworkstyle idle unless workspace titles actually need updating (and should make it slightly more responsive at the same time):

- Instead of waking up ten times per second to poll for Sway events and config changes, it uses async-io to set up a combined stream of both. This keeps the sworkstyle process asleep until it has work to do, and respond slightly faster when it does (because it would previously have to wait out the 100ms sleep before responding).
- Instead of refreshing window titles after all window events, it skips a few irrelevant ones, particularly focus changes. It still receives these (so it still has to wake up) but it no longer gets the full window tree from Sway as often.
- It updates the workspace titles once on startup and after config reloads. Necessary because it now takes more than just a focus change for these to take effect.

It also includes some housekeeping. The edition update is not strictly necessary: I originally expected to use an async closure (which requires the 2024 edition) but changed my mind, and dropped the 2024 edition upgrade but kept the 2021 one. The addition of rustfmt.toml specifying the 2018 edition or newer is desirable because these changes introduce an async block, which is invalid syntax before that edition: without rustfmt.toml editors that invoke rustfmt directly will fail.

marienz committed:
Update dependencies

inotify 0.10.1 deprecated add_watch. Use the new API to avoid a compiler
warning.

The other changes (including backwards-incompatible changes to
simple_logger, toml and dirs) do not affect us.

marienz committed:
Upgrade to edition 2021 (no changes)

marienz committed:
cargo fmt (2021 edition)

marienz committed:
Explicitly specify edition for rustfmt

This makes sure automated formatting works when not invoked through
Cargo as we start using features require a more recent edition.

marienz committed:
Clarify link between inotify and config_path

inotify and config_path are both options, but it is not possible for
inotify to be set without config_path also being set.

Make this explicit by adding a struct.

This does not make much of a difference now, but helps with further
changes.

marienz committed:
Do not wake up unnecessarily

To simultaneously check for window events from Sway and config updates,
we were polling (check both, then sleep 100ms, in a loop). This means we
always wake up ten times per second, which is inefficient: although the
CPU usage should not be too noticeable on a modern CPU, it's not great
for power efficiency on an otherwise idle system.

Improve matters by asynchronously checking for config updates:

- Switch from the async-std and futures crates to async-io and
  futures-lite. These are what swayipc-async uses under the hood,
  and they provide all functionality we need: we might as well use them.
- Use async-io to poll inotify asynchronously, producing a stream of
  config updates.
- Merge this with the stream of events we already have, and process both
  in parallel.

This results in sworkstyle being completely idle between window
events (and it should respond slightly faster to them as it no longer
needs to wait an average 50ms in the 100ms sleep between events).

marienz committed:
Refresh workspace names when relevant

Ignore some Sway events that should not affect our workspace titles,
particularly focus changes.

Refresh workspace titles immediately after a config reload. This was
previously not as necessary because a focus change would force an
update.

marienz committed:
Update workspace names once on startup

Like updating after config reload, this matters more because we no
longer update on focus change.

Lyr-7D1h commented on src/lib.rs:
```diff
@@ -20,81 +19,120 @@ use config::Config;
-                if let Ok(_) = inotify.read_events(&mut inotify_events_buffer) {
-                    if let Some(config_path) = &self.config_path {
+        if let Some(source) = self.config_source.take() {
+            if source.inotify.is_some() {
```
A bit of a nitpick but its cleaner to replace this with `if let Some(inotify) = source.inotify` as you are unwrapping later.

marienz replied:
Possible with some further changes to the function passed to `stream::try_unfold`. Can you please check the commit I've added and confirm that's what you had in mind? I had it like this before but I found the tuples a little unwieldy, but you're probably right that the unwrap() is worse.

(Replacing the tuple with a struct containing a non-optional `Inotify` would also work: I'm tempted to replace `ConfigSource` with that as `ConfigSource` without `Inotify` is currently useless. That'd mean if `Sworkstyle::new` can't init inotify you'd get a `Sworkstyle` that contains a config, but not the path it was loaded from. What do you think?)

(It's all slightly fiddly because `Async::new` takes ownership of `inotify`, which means the async closure needs to take ownership, pass it to `Async::new`, and then transfer ownership back... although the `inotify` we get back from `anotify.into_inner()` is the same thing the closure was originally called with, Rust's type system does not know that, so everything I've tried to make this simpler results in "value used after move" errors. Just closing over `inotify` looks like it should work but doesn't because of that ownership back-and-forth.)

Lyr-7D1h replied:
Yes I agree, having ConfigSource without Inotify doesn't make that much sense. The source should always be a valid path anyways. 

So creating a config source when a config path is given with a non optional inotify would be ideal

marienz replied:
Ok, how about something like this?

(Reasonably happy with how it looks myself but also happy to tweak it further if you have suggestions, my motivation for sending the PR was about an even split between "reduce wakeups" and "tinker with some code" anyway :)

Lyr-7D1h commented:
Thank you so much for this well considered PR! Just one small comment but then otherwise ready to be merged and put in version 1.3.7

marienz committed:
Update expected TOML error

An update of the TOML crate changed the error message.

Tweak the test to log the error we actually got to make this easier to
fix in the future.

marienz committed:
Avoid unnecessary unwrap

marienz commented:
Turns out I never ran `cargo test`, and missed I'd broken that with the dependency update (whooooopppps).

The "Update expected TOML error" commit I just added should get squashed into "Update dependencies" commit. I haven't done that yet because I don't know if Github's review workflow will handle the following rebase well. I can take care of that after you've reviewed the further changes.

Lyr-7D1h commented:
I wouldn't worry too much about keeping a clean commit history as this is a small project.

marienz committed:
Refactor ConfigSource to require Inotify

We have no current or planned use for the config path if it did not
exist at startup, so there is no point in keeping it around.

Although it's not much shorter, this removes two levels of indentation
from the `stream::try_unfold` call (by removing an `if` and simplifying
the signature of the closure).

marienz commented:
Any further changes you'd like for this before it's merged?
